### PR TITLE
Implement #160 Related Origins support (WebAuthn L3 § 5.11)

### PR DESCRIPTION
**Pull request by HarveyOrourke15**
This pull request introduces a new implementation for validating related origins in accordance with the WebAuthn specification (§5.11.1). 

* Added a new `related_origins` module to `libwebauthn`.
* Provided a public API (`validate_related_origins`) and a pluggable HTTP client trait (`RelatedOriginsHttpClient`) for extensible validation and document fetching.
* `ReqwestRelatedOriginsClient` using the `reqwest` crate, enforces HTTPS, content type, content length, and streaming limits for secure fetching of well-known documents.
* Added new dependencies: `reqwest`, `bytes`, `url`, and `publicsuffix` to `Cargo.toml` to support HTTP requests, streaming, URL parsing, and domain label extraction.

**Testing and examples:**

* Added unit and property tests for the validator and domain label extraction, covering edge cases and policy limits.
* Introduced a new example (`debug_related_origins.rs`) demonstrating environment-driven validation calls for debugging and manual testing.
* Added an integration test for Amazon's related origins validation.

### CODE REVIEW

File: libwebauthn/src/ops/webauthn/related_origins/client.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+impl RelatedOriginsHttpClient for ReqwestRelatedOriginsClient {
+    async fn fetch_well_known(&self, rp_id: &str) -> Result<RelatedOriginsHttpResponse, RelatedOriginsError> {
+        let url = format!("https://{}/.well-known/webauthn", rp_id);
+
+        // When issuing the request, avoid cookies/auth and set no Referer:
+        let req = self.inner.get(&url).header(reqwest::header::REFERER, "");
+        let resp = req.send().await.map_err(|e| RelatedOriginsError::HttpError(e.to_string()))?;
+
+        let status = resp.status();
+        if status != StatusCode::OK {
```
Comment (msirringhaus) [inline libwebauthn/src/ops/webauthn/related_origins/client.rs:46]: All these checks happen outside in `validate_related_origins_with_client()` as well (status code, content type, content length, ..). Is there a reason to duplicate this here?

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+use futures::TryStreamExt;
+
+/// Central policy/config for related-origins processing. Consumers may
+/// construct a custom `PolicyConfig` and use it when building a client.
+#[derive(Clone, Debug)]
+struct PolicyConfig {
+    pub max_body_bytes: usize,
+    pub max_origins: usize,
+    pub max_origin_len: usize,
+    pub max_labels: usize,
```
[the remark this answers is not in the document]
Comment (msirringhaus) [reply libwebauthn/src/ops/webauthn/related_origins/mod.rs:19]: Correct.

File: libwebauthn/src/ops/webauthn/related_origins/client.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    }
+}
+
+#[async_trait::async_trait]
+impl RelatedOriginsHttpClient for ReqwestRelatedOriginsClient {
+    async fn fetch_well_known(&self, rp_id: &str) -> Result<RelatedOriginsHttpResponse, RelatedOriginsError> {
+        let url = format!("https://{}/.well-known/webauthn", rp_id);
+
+        // When issuing the request, avoid cookies/auth and set no Referer:
+        let req = self.inner.get(&url).header(reqwest::header::REFERER, "");
```
[the remark this answers is not in the document]
Comment (msirringhaus) [reply libwebauthn/src/ops/webauthn/related_origins/client.rs:42]: additionally: Since this is something all clients would need to do, I wonder if we shouldn't just do this on the outside and pass the final well-known url in here.

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+) -> Result<bool, RelatedOriginsError>
+where
+    C: RelatedOriginsHttpClient + Sync,
+{
+    // Use default policy for validation.
+    let policy = PolicyConfig::default();
+
+    // Fetch the well-known URL for the RP ID
+    let response = client.fetch_well_known(rp_id_requested).await.map_err(|_| RelatedOriginsError::SecurityError)?;
+
```
[the remark this answers is not in the document]
Comment (msirringhaus) [reply libwebauthn/src/ops/webauthn/related_origins/mod.rs:82]: Agreed, `tracing` the error details would be good.

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    if response.status != 200 {
+        return Err(RelatedOriginsError::SecurityError);
```
Comment (msirringhaus) [inline libwebauthn/src/ops/webauthn/related_origins/mod.rs:84]: `tracing`-calls here and below, for each early return would be good, to pinpoint errors quickly.

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+/// delegates to `validate_related_origins_with_client`.
+pub async fn validate_related_origins(
+    caller_origin: &str,
+    rp_id_requested: &str,
+) -> Result<bool, RelatedOriginsError> {
+    let client = match client::ReqwestRelatedOriginsClient::new() {
+        Ok(c) => c,
+        Err(e) => {
+            eprintln!("related_origins: failed to create HTTP client: {:?}", e);
+            return Err(e);
```
[the remark this answers is not in the document]
Comment (msirringhaus) [reply libwebauthn/src/ops/webauthn/related_origins/mod.rs:164]: Correct. Please replace `eprintln!()` with `tracing`.

File: libwebauthn/Cargo.toml [hunk 7]
```diff
@@ -73,6 +73,10 @@ ctap-types = { version = "0.4.0" }
 btleplug = "0.11.7"
 thiserror = "2.0.12"
 serde_json = "1.0.141"
+url = "2.4.0"
+publicsuffix = "1.5"
+reqwest = { version = "0.11", features = ["rustls-tls", "stream"] }
```
[the remark this answers is not in the document]
Comment (msirringhaus) [reply libwebauthn/Cargo.toml:78]: I don't think that is true for newer `reqwest`-versions. Somebody would need to verify that.

Comment (msirringhaus) [inline libwebauthn/Cargo.toml:78]: This needs to be behind a feature flag. The issue suggested two feature flags:
```
[features]
related-origins = []
related-origins-client = ["related-origins", "reqwest"]   # I would call this `related-origins-client-reqwest`, in case we get more implementations/backends
```
I think, for now I could also live with one flag, that encompasses all added dependencies here. But we have to be able to exclude at least `reqwest`, the example binary and the whole `related_origins`-module.

File: libwebauthn/tests/related_origins.rs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+use libwebauthn::ops::webauthn::related_origins::validate_related_origins;
+
+#[tokio::test]
+async fn amazon_validate_related_origins() -> Result<(), Box<dyn std::error::Error>> {
+    let caller_origin = "https://www.amazon.com";
+    let rp_id = "amazon.com";
+
+    let res = validate_related_origins(caller_origin, rp_id).await?;
```
[the remark this answers is not in the document]
Comment (msirringhaus) [reply libwebauthn/tests/related_origins.rs:8]: Hm, if we simply mark it `#[ignore]`, it will just never get called/tested. I'm not sure how often it would actually fail in CI. @AlfioEmanueleFresta thoughts?

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+        }
+    };
+
+    validate_related_origins_with_client(caller_origin, rp_id_requested, &client).await
+}
+
+// Client implementation is kept in the `client` submodule.
+pub mod client;
```
Comment (msirringhaus) [inline libwebauthn/src/ops/webauthn/related_origins/mod.rs:172]: This needs to be behind a feature flag, as mentioned in the comment of `Cargo.toml`.

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+/// Implimentation of https://url.spec.whatwg.org/#host-public-suffix
```
[the remark this answers is not in the document]
Comment (msirringhaus) [reply libwebauthn/src/ops/webauthn/related_origins/mod.rs:174]: Correct.

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+    }
+
+    // Tests use the public API and the `registrable_origin_label` helper.
+
+    #[tokio::test]
+    async fn test_validate_related_origins_true() {
+        let doc = r#"{"origins":["https://example.co.uk","https://example.de"]}"#;
+        let client = MockClient { body: doc.into() };
+        let res = validate_related_origins_with_client("https://example.co.uk", "example.com", &client).await.unwrap();
+        assert!(res);
```
Comment (msirringhaus) [inline libwebauthn/src/ops/webauthn/related_origins/mod.rs:235]: nit: I find it more readable to remove the `unwrap()` and do `assert_eq!(res, Ok(true))`.

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+    }
+
+    #[tokio::test]
+    async fn test_validate_related_origins_false() {
+        let doc = r#"{"origins":["https://different.example.co.uk"]}"#;
+        let client = MockClient { body: doc.into() };
+        let res = validate_related_origins_with_client("https://caller.example.com", "example.com", &client).await.unwrap();
+        assert!(!res);
```
Comment (msirringhaus) [inline libwebauthn/src/ops/webauthn/related_origins/mod.rs:243]: even more so here

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+        assert_eq!(registrable_origin_label("localhost"), None);
+        assert_eq!(registrable_origin_label("com"), None);
+    }
+
+    #[tokio::test]
+    async fn test_validate_empty_origins() {
+        let doc = r#"{"origins":[]}"#;
+        let client = MockClient { body: doc.into() };
+        let res = validate_related_origins_with_client("https://caller.example.com", "example.com", &client).await;
+        match res {
```
Comment (msirringhaus) [inline libwebauthn/src/ops/webauthn/related_origins/mod.rs:259]: `assert_eq!(res, Err(RelatedOriginsError::SecurityError))` works here as well and is far shorter than the `match`. (One can also add additional output after, like `assert_eq!(res, Err(..), "Needs to be SecurityError, because of..")`.)

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+            other => panic!("expected SecurityError, got {:?}", other),
+        }
+    }
+
+    #[tokio::test]
+    async fn test_validate_non_https_origin() {
+        let doc = r#"{"origins":["http://example.com"]}"#;
+        let client = MockClient { body: doc.into() };
+        let res = validate_related_origins_with_client("https://caller.example.com", "example.com", &client).await;
+        match res {
```
Comment (msirringhaus) [inline libwebauthn/src/ops/webauthn/related_origins/mod.rs:270]: Same here. `assert_eq!(res, Err(RelatedOriginsError::SecurityError), "expected SecurityError for non-https origin, got {:?}", res)` should do the trick

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+
+    #[tokio::test]
+    async fn test_validate_origin_too_long() {
+        // create a body with an origin string longer than the default max_origin_len
+        let long_path = "a".repeat(2100);
+        let origin = format!("https://example.com/{}", long_path);
+        let doc = serde_json::json!({"origins": [origin]}).to_string();
+        let client = MockClient { body: doc };
+        let res = validate_related_origins_with_client("https://caller.example.com", "example.com", &client).await;
+        match res {
```
Comment (msirringhaus) [inline libwebauthn/src/ops/webauthn/related_origins/mod.rs:284]: dito

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+    // Host must look like a domain (contain at least one dot)
+    if !host.contains('.') {
+        return None;
+    }
+
+    if let Some(list_opt) = PS_LIST.get_or_init(|| publicsuffix::List::fetch().ok()) {
+        if let Ok(parsed) = list_opt.parse_domain(host) {
+            // If the parsed result provides the registrable root (eTLD+1),
+            // take the left-most label of that root as the registrable label.
+            if let Some(root) = parsed.root() {
```
[the remark this answers is not in the document]
Comment (msirringhaus) [reply libwebauthn/src/ops/webauthn/related_origins/mod.rs:199]: We had this discussion before, with no clear solution. The list changes too often to rely only on compile time-snapshots. A hybrid approach would be thinkable, where we use both, try to fetch an updated version, and if that fails, fall back to the compile-time list. 
But that would also increase the dependency tree and general bloaty-ness.
@AlfioEmanueleFresta , do you have a strong opinion on this either way?

**Review by msirringhaus — CHANGES_REQUESTED** [review]
Thanks for this! Looking good overall!
I left comments to some minor things

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (AlfioEmanueleFresta) [reply libwebauthn/src/ops/webauthn/related_origins/mod.rs:199]: Fetching a new list upon each request is expensive. A very outdated PSL is a security risk. I don't think we should try and manage a PSL ourselves if we can at all avoid it.

I think the ideal solution would be to use the system-managed PSL in `/usr/share/publicsuffix`. From a quick search, it seems to be widely available on most Linux distros. It's updated via package managers. We could fallback to downloading our own, but this should be cached for a reasonable time. I'm not sure what time time should be.

A compromise might be that of using a package like [public-suffix](https://github.com/1Password/passkey-rs/tree/main/public-suffix), which ships with its own PSL. However, it doesn't seem to be updated very frequently (last update 10 months ago) and it also relies on the version of libwebauthn and publix-suffix in the chain of dependencies, which can get obsolete quickly.

WDYT?

File: libwebauthn/tests/related_origins.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlfioEmanueleFresta) [reply libwebauthn/tests/related_origins.rs:8]: I don't think our CI has network access other than crates.io (or, I hope it doesn't :)).

I would rather we build a test implementation of the HTTP client  trait, and provide a mock list. This would allow tests to run automatically in CI, without an external dependency. 

We should also avoid using real domain names in favour of reserved example domains.

File: libwebauthn/src/ops/webauthn/related_origins/mod.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (iinuwa) [reply libwebauthn/src/ops/webauthn/related_origins/mod.rs:199]: > I think the ideal solution would be to use the system-managed PSL in /usr/share/publicsuffix. From a quick search, it seems to be widely available on most Linux distros. It's updated via package managers. We could fallback to downloading our own, but this should be cached for a reasonable time. I'm not sure what time time should be.

I think using this package makes sense, but I think reading this file belongs in the credentialsd layer, not `libwebauthn`. For example, if someone wanted to build a browser, they probably already have a method to determine PSLs, so they would want to leverage their own infrastructure rather than requiring ours. In other words, I think libwebauthn could:
- accept the public suffix list in a specific format, or
- use a callback trait to allow the caller to verify a specific domain,

and leave it up to callers to provide the PSL verification. If necessary, we could provide a default PSL implementation under a non-default feature flag as a convenience.

[the remark this answers is not in the document]
Comment (AlfioEmanueleFresta) [reply libwebauthn/src/ops/webauthn/related_origins/mod.rs:199]: I agree with @iinuwa:
* The client should be able to override PSL behaviour.
* I am also keen for `libwebauthn` to provide a safe, and implementation for the PSL.

I think a good compromise would be to:
1. Define a `PublicSuffixList` trait, with a single `async` method to get the registrable domain as needed by the spec;
2. Have `libwebauthn` include a simple implementation, based on `publicsuffix` reading a PSL from a static DAT file path.

E.g.: 
```rust
#[async_trait]
pub trait PublicSuffixList: Send + Sync {
    /// Returns the registrable domain (eTLD+1) for the given host, if any.
    async fn registrable_domain(&self, host: &str) -> Option<String>;
}
```

And:

```rust
pub struct DatFilePublicSuffixList {
  fn new(path: &PathBuf) -> Option<Self> { ... }
  fn from_system_file() -> Option<Self> { 
    Self::new(PathBuf::from("/usr/share/publicsuffix/public_suffix_list.dat"))
  }
}
```

Clients like credentialsd can then decide whether to:
* (a) Depend on the system package which provides the DAT file at a known location, using libwebauthn's verifier with the appropriate file path, or
* (b) to provide theyr own implementation, which downloads & caches a PSL file

As for this specific PR, I think it would be fine just to implement and use the `DatFilePublicSuffixList` DAT-based PSL verifier component. 

Building a `PublicSuffixList` trait, and allowing specifying/providing a PSL can be done as a follow up - by the same author or someone else.

WDYT?

**Review by AlfioEmanueleFresta — CHANGES_REQUESTED** [review]
Thank you for all your work so far on this @HarveyOrourke15!

**Comment by AlfioEmanueleFresta** [comment]
Thank you for your work on this @HarveyOrourke15! As it'd been a while I continued the work in #219, so I'm closing this PR.
